### PR TITLE
Airfield Runway Crosswind

### PR DESCRIPTION
**Pull request by norecords**
This is a special feature for Airfield Weather Stations,
it calculate the Headwind speed, the Crosswind speed and direction from an active Runway number that is set in skin.conf.
The Active Runway depend of the wind direction and will switch to his opposite to be always in front of the wind.
That show the lane with active Runway number and is opposite on the compass.
From the initiative of @flygpagen from issue #596

**Actually works only with Mqtt enabled and Mph, Kph, Mps or Knot windspeed.**

### CODE REVIEW

**Comment by pgpaolo** [comment]
Hello,
thanks for your implementation, it is very useful.

MQTT active - does not work

MQTT off - does not work

what do you think about it?

Thanks

http://meteostz05013.ddns.net/

**Comment by norecords** [comment]
@pgpaolo I know that my screenshots can make confused, but as I wrote on the first comment of this PR, this actually only works with **Knot** speed.
I'll write something in the next days to be compatible with all kind of wind speed.
As you use km/h, you can try this quick hack:
Find and edit belchertown.js.tmpl:
```
function calcXwind(windDir, windspeed, runway) {
    windDir = parseFloat(windDir).toFixed(0);
    // Find Active Runway relative to the wind direction
    runway = activeRunway(windDir,runway * 10);
    // Windspeed must be Knots
    windspeed = Math.round(parseFloat(windspeed).toFixed(0) / 1.852); // km/h to knots
    var lrxw = "";
    belchertown_debug("windDir: " + windDir + " Windspeed (knots): " + windspeed + " Runway Heading (°): " + runway);
    let u = Math.abs(windDir - runway); // Crosswind angle
    const crosswind = Math.round(Math.abs(windspeed * Math.sin(0.017453 * u))),
          headwind  = Math.round(windspeed * Math.cos(0.017453 * u));
    u > 180 && (u = Math.abs(u - 360));
    0 === u || 180 === u ? (lrxw = "$obs.label.airfield_crosswind_none") : (windDir > runway && windDir < runway + 180) || runway - 180 > windDir ? (lrxw = "$obs.label.airfield_crosswind_right") : (windDir < runway || windDir > runway + 180) && (lrxw = "$obs.label.airfield_crosswind_left");
    belchertown_debug("Headwind: " + headwind + " Crosswind: " + crosswind + " Xwind Dir: " + lrxw + " Xwind angle: " + u);
    jQuery(".active-runway").html(runwayNumber(runway / 10));
    jQuery(".headwind").html(Math.round(headwind*1.852)); // Knots to km/h
    jQuery(".crosswind").html(Math.round(crosswind*1.852));
    jQuery(".crosswind-lrxw").html(lrxw);
}
```
and @ line 1910:
```
    #if $Extras.has_key("airfield_crosswind_enabled") and $Extras.airfield_crosswind_enabled == '1'
    // Airfield Crosswind
    if (data.hasOwnProperty("windDir") && data.hasOwnProperty("windSpeed_kph")) {
        calcXwind(data["windDir"], data["windSpeed_kph"], active_runway);
    }
    #end if
```

**Comment by norecords** [comment]
@pgpaolo you also have a mistake in index.html.tmpl:
find
`ranway-south` and change to `runway-south`

**Comment by pgpaolo** [comment]
Thanks for your support.
I was wrong wind speed knots to kph.

and so many stupid mistakes. :(
I made the changes on the fly :)
Now it works

GP

**Comment by norecords** [comment]
Good to hear, I can see that you still have missing commits, look at https://github.com/poblabs/weewx-belchertown/pull/657/files and compare to your changes.

**Comment by pgpaolo** [comment]
> Good to hear, I can see that you still have missing commits, look at https://github.com/poblabs/weewx-belchertown/pull/657/files and compare to your changes.

Hi,
unfortunately, the real-time beafort  scale no longer works

What do you think it?

Thanks

GP

**Review by flygpagen — APPROVED** [review]
Enables feature of showing a runway in the wind compass, usable for weather stations located at airfields with 1 runway. Also displays current runway in use, headwind and crosswind with labels depending on left or right crosswind.

File: skins/Belchertown/js/belchertown.js.tmpl [hunk 1]
```diff
@@ -386,6 +397,34 @@ function beaufort_cat(beaufort) {
     }
 }
 
+#if $Extras.has_key("airfield_crosswind_enabled") and $Extras.airfield_crosswind_enabled == '1'
+const active_runway = parseFloat($Extras.airfield_crosswind_active_runway).toFixed(0);
+
+// Find Opposite Runway
+var opposite_runway = (((active_runway * 10) + 180) % 360) / 10;
```
Comment (dimme) [inline skins/Belchertown/js/belchertown.js.tmpl]: Or simply:
var opposite_runway = (active_runway + 18) % 36;

File: skins/Belchertown/js/belchertown.js.tmpl [hunk 2]
```diff
@@ -386,6 +397,34 @@ function beaufort_cat(beaufort) {
...
+
+// Positionning Active Runway and Opposite at the right place
+var css_north = opposite_runway;
+var css_south = active_runway;
+
+function calcXwind(windDir, windspeed, runway) {
```
Comment (dimme) [inline skins/Belchertown/js/belchertown.js.tmpl:417]: I don't know if your function was always wrong but for wind 360/10 on RWY 36, it gives the wrong result.

Here is my implementation, read the comments for more:

`function calcXwind(winddir, windspeed, runway) {
    // Use float values as that will provide better precision with the triginometric functions below.
    winddir = parseFloat(winddir);
    windspeed = parseFloat(windspeed);
    
    // The angle alpha is the difference between the wind and the active runway, in degrees. Use (Math.PI / 180) to convert to radians more precise.
    const alpha = (winddir - runway * 10) * (Math.PI / 180);

    // If the headwind becomes negative, then you have not fed the active runway into the function.
    const headwind = Math.round(windspeed * Math.cos(alpha));

    // Positive crosswind is from the right, negative is from the left.
    const crosswind = Math.round(windspeed * Math.sin(alpha));
    const direction = crosswind == 0 ? 'None' : crosswind > 0 ? 'Right' : 'Left';

    // Show the values
    jQuery(".headwind").html(headwind);
    jQuery(".crosswind").html(Math.abs(crosswind));
    jQuery(".crosswind_lrxw").html(direction);
}`

**Review by dimme — CHANGES_REQUESTED** [review]
See my comments.

File: skins/Belchertown/js/belchertown.js.tmpl (hunk 2 above)
Comment (norecords) [reply skins/Belchertown/js/belchertown.js.tmpl:417]: Thank you for the interest in this modification,
I had not seen this error before, and will try your code snippet / optimization asap.
On other part I'm working on a better css / html integration.

**Comment by norecords** [comment]
> Hi,
> unfortunately, the real-time beafort scale no longer works
> 
> What do you think it?
> 
> Thanks
> 
> GP

It haven't been tested with that option enabled, so I can't tell you right now

File: skins/Belchertown/js/belchertown.js.tmpl (hunk 2 above)
Comment (norecords) [reply skins/Belchertown/js/belchertown.js.tmpl:417]: I'm comparing results between your function and mine and have same final results so I can't see where the error. Any way your code is cleanest, faster, and more precise.

**Comment by pgpaolo** [comment]
> > Ciao,  sfortunatamente, la scala beafort in tempo reale non funziona più
> > Cosa ne pensi?
> > grazie
> > Gp
> 
> @pgpaolo Non è stato testato con quell'opzione abilitata, quindi non posso dirtelo in questo momento

@norecors Hi,problem solved 
The error was in this function 

belchertown.js.tmpl
#if $Extras.has_key("beaufort_category") and $Extras.beaufort_category == '1'
    if (data.hasOwnProperty("beaufort_count")) {
        jQuery(".beaufort").html(beaufort_cat(parseFloat(data["beaufort_count"])));
        belchertown_debug("Wind speed: " + parseFloat(data["beaufort_count"]) + " Beaufort cat: " + beaufort_cat(parseFloat(data["**beaufort_count**"])))

before:
beaufort

after:
beaufort_count

Thanks 
:)

File: skins/Belchertown/js/belchertown.js.tmpl (hunk 2 above)
Comment (norecords) [reply skins/Belchertown/js/belchertown.js.tmpl]: Changes merged https://github.com/poblabs/weewx-belchertown/pull/657/commits/597485f2dd1c82cf7c94dd435309e091b34e1355

**Comment by norecords** [comment]
Thanks for your help @dimme I add your change in last commit, can you give a try.

@flygpagen @pgpaolo can you also test this commit https://github.com/poblabs/weewx-belchertown/pull/657/commits/597485f2dd1c82cf7c94dd435309e091b34e1355

**Comment by pgpaolo** [comment]
@norecords Hi, 
first code:
// Active Runway from skin.conf ----- ok
// Find Opposite Runway ---------ok
function activeRunway --------- ok 

calcXwind not function

the code I use makes a modification to calculate the wind speed in knots: 
        **windspeed_h = (parseFloat(windspeed).toFixed(1) / 1.852); // km/h to knots
	windspeed = Math.round(parseFloat(windspeed).toFixed(2) / 1.852); // km/h to knots
	jQuery(".air_knots2").html(parseFloat(windspeed_h).toFixed(1));**



function calcXwind(windDir, windspeed, runway) {
    windDir = parseFloat(windDir).toFixed(0);
    // Find Active Runway relative to the wind direction
    runway = activeRunway(windDir,runway * 10);
    // Windspeed must be Knots
    windspeed_h = (parseFloat(windspeed).toFixed(1) / 1.852); // km/h to knots
	windspeed = Math.round(parseFloat(windspeed).toFixed(2) / 1.852); // km/h to knots
    var lrxw = "";
    belchertown_debug("windDir: " + windDir + " Windspeed (knots): " + windspeed + " Runway Heading (°): " + runway);
    let u = Math.abs(windDir - runway); // Crosswind angle
    const crosswind = Math.round(Math.abs(windspeed * Math.sin(0.017453 * u))),
          headwind  = Math.round(windspeed * Math.cos(0.017453 * u));
		  u > 180 && (u = Math.abs(u - 360));
    0 === u || 180 === u ? (lrxw = "$obs.label.airfield_crosswind_none") : (windDir > runway && windDir < runway + 180) || runway - 180 > windDir ? (lrxw = "$obs.label.airfield_crosswind_right") : (windDir < runway || windDir > runway + 180) && (lrxw = "$obs.label.airfield_crosswind_left");
    belchertown_debug("Headwind: " + headwind + " Crosswind: " + crosswind + " Xwind Dir: " + lrxw + " Xwind angle: " + u);
    jQuery(".active-runway").html(runwayNumber(runway / 10));
    jQuery(".headwind").html(Math.round(headwind*1.852)); // Knots to km/h
    jQuery(".crosswind").html(Math.round(crosswind*1.852));
    jQuery(".crosswind-lrxw").html(lrxw);
	
	jQuery(".air_knots2").html(parseFloat(windspeed_h).toFixed(1));


Second code:
calcXwind(parseFloat(data["windDir"]), parseFloat(data["windSpeed_knot"]), activeRunway(parseFloat(data["windDir"]).toFixed(0), active_runway));

if I insert the second part together with the first by replacing knot to kph.

Crash :(

thanks

GP

**Comment by norecords** [comment]
@pgpaolo if you don't make changes as committed, I can't help you.
calcXwind should look like this
```
function calcXwind(winddir, windspeed, runway) {
    // The angle alpha is the difference between the wind and the active runway, in degrees. Use (Math.PI / 180) to convert to radians more precise.
    const alpha = (winddir - runway * 10) * (Math.PI / 180);

    // Calculate the headwind.
    const headwind = Math.round(windspeed * Math.cos(alpha));

    // Positive crosswind is from the right, negative is from the left. Become none when wind is come from front or no wind speed.
    const crosswind = Math.round(windspeed * Math.sin(alpha));
    const direction = crosswind == 0 ? "$obs.label.airfield_crosswind_none" : crosswind > 0 ? "$obs.label.airfield_crosswind_right" : "$obs.label.airfield_crosswind_left";

    // Show the values
    jQuery(".headwind").html(headwind);
    jQuery(".crosswind").html(Math.abs(crosswind));
    jQuery(".crosswind-lrxw").html(direction);

}
```

**Comment by pgpaolo** [comment]
@norecords 
debug enabled
i put all the new code

Does not work
thx

**Comment by norecords** [comment]
@pgpaolo Please re-read your changes...
`calcXwind(parseFloat(data["windDir"]), parseFloat(data["windSpeed_knot"]), activeRunway(parseFloat(data["windDir"]).toFixed(0), active_runway));`
should be
`calcXwind(parseFloat(data["windDir"]), parseFloat(data["windSpeed_kph"]), activeRunway(parseFloat(data["windDir"]).toFixed(0), active_runway));`
to fit your needs.

**Comment by pgpaolo** [comment]
> @pgpaolo Ri-leggere le modifiche... dovrebbe
> 
> essere quello di soddisfare le vostre esigenze.
> 
> `calcXwind(parseFloat(data["windDir"]), parseFloat(data["windSpeed_knot"]), activeRunway(parseFloat(data["windDir"]).toFixed(0), active_runway));

``calcXwind(parseFloat(data["windDir"]), parseFloat(data["windSpeed_kph"]), activeRunway(parseFloat(data["windDir"]).toFixed(0), active_runway));`

@norecords 
doesn't work with both  -kph -knot

**Comment by norecords** [comment]
@dimme is the wind speed unit care in the calculation, this still need to be knot ?

**Comment by dimme** [comment]
> @dimme is the wind speed unit care in the calculation, this still need to be knot 

The wind speed unit can be anything. If you feed in knots you get out knots. If you feed in km/h you get out km/h.

**Comment by norecords** [comment]
This have been tested on a fresh install and works as expected.

**Review by norecords — COMMENTED** [review]
Added into the script.

**Comment by fvirgola80** [comment]
> Thanks for your support.
> I was wrong wind speed knots to kph.
> 
> and so many stupid mistakes. :(
> I made the changes on the fly :)
> Now it works
> 
> GP

Hi @pgpaolo, I took a look at your site and I would like to know how did you integrate the wind speed also in knots. I would like to do it too, could you help me with this please? thanks my email is fvirgola80@gmail.com while my website is [www.meteocivitavecchia.it](url)
Thanks

**Comment by norecords** [comment]
@fvirgola80 could you please open a new issue and mention @pgpaolo in there. Thank you

**Comment by fvirgola80** [comment]
> @fvirgola80 could you please open a new issue and mention @pgpaolo in there. Thank you

ok thank you :-)

**Comment by pgpaolo** [comment]
> Added into the script:
> 
> * Works with mph, kph, mps and knot wind speed.
> * Bold active runway on compass.
> * Correct css and html syntax.
> * Corrected parentheses in activeRunway function.
> 
> Screenshot_2021-04-06 Météo Correns - Station météorologique en temps réel
> Screenshot_2021-04-06 Météo Correns - Station météorologique en temps réel(1)

@norecords 
Thanks
positive feedback

now it works correctly.
it's faster

**Review by flygpagen — APPROVED** [review]

**Review by flygpagen — APPROVED** [review]

**Comment by norecords** [comment]
@flygpagen @pgpaolo thanks for your feedback.

**Comment by flygpagen** [comment]
Suggestion for docs/readme;

> Enables feature of showing a runway in the wind compass, usable for weather stations located at airfields with 1 runway. Also displays current runway in use, headwind and crosswind with labels depending on left or right crosswind. Add the runway heading (in degrees, 22 for 220 degrees) in the skins.conf and enable feature, then wait for Weewx to generate HTML to see the results.

**Comment by norecords** [comment]
@flygpagen because you use it permanently, can we add a link to your station as a live demo ?

**Comment by flygpagen** [comment]
> @flygpagen because you use it permanently, can we add a link to your station as a live demo ?

Yes, no problem! :)

**Pull request closed without merging**